File: backend/app/services/prices.py
"""Live prices.

Mutual funds : AMFI publishes all NAVs daily as a semicolon-delimited text
               file. It includes ISINs, so NSDL/CDSL CAS holdings (which key
               on ISIN) map cleanly.
Stocks       : yfinance by NSE/BSE ticker. ISIN -> ticker is resolved via a
               bundled mapping file (data/isin_symbols.csv) which you can
               extend; unresolved ISINs are reported so the user can map them
               in the UI.
"""
from __future__ import annotations

import csv
import logging
import time
from datetime import date, datetime
from pathlib import Path

import httpx

log = logging.getLogger(__name__)

AMFI_NAV_URL = "https://portal.amfiindia.com/spages/NAVAll.txt"
DATA_DIR = Path(__file__).resolve().parents[2] / "data"

_amfi_cache: dict = {"fetched_at": 0.0, "by_isin": {}}
AMFI_TTL_SECONDS = 6 * 3600


async def fetch_amfi_navs(force: bool = False) -> dict[str, dict]:
    """Return {isin: {nav, name, amfi_code, nav_date}} for every scheme."""
    now = time.time()
    if not force and _amfi_cache["by_isin"] and now - _amfi_cache["fetched_at"] < AMFI_TTL_SECONDS:
        return _amfi_cache["by_isin"]

    async with httpx.AsyncClient(timeout=60, follow_redirects=True) as client:
        resp = await client.get(AMFI_NAV_URL)
        resp.raise_for_status()

    by_isin: dict[str, dict] = {}
    # Format: Scheme Code;ISIN Div Payout/Growth;ISIN Div Reinvestment;Scheme Name;NAV;Date
    for line in resp.text.splitlines():
        parts = line.split(";")
        if len(parts) != 6 or not parts[0].strip().isdigit():
            continue
        code, isin1, isin2, name, nav_s, date_s = [p.strip() for p in parts]
        try:
            nav = float(nav_s)
            nav_date = datetime.strptime(date_s, "%d-%b-%Y").date()
        except ValueError:
            continue
        rec = {"nav": nav, "name": name, "amfi_code": code, "nav_date": nav_date}
        for isin in (isin1, isin2):
            if isin and isin != "-":
                by_isin[isin] = rec

    _amfi_cache.update(fetched_at=now, by_isin=by_isin)
    log.info("AMFI: cached %d ISINs", len(by_isin))
    return by_isin


def _yf_sym(sym: str) -> str:
    """Normalise Indian ticker for yfinance: add .NS if missing and not BSE."""
    if "." not in sym:
        return sym + ".NS"
    return sym


def fetch_stock_prices(symbols: list[str]) -> dict[str, dict]:
    """Return {symbol: {price, price_date}} via yfinance. Sync (yfinance is)."""
    if not symbols:
        return {}
    import yfinance as yf  # local import: heavy

    # Normalise symbols for NSE (yfinance requires .NS suffix)
    sym_map = {s: _yf_sym(s) for s in symbols}
    yf_symbols = list(set(sym_map.values()))

    out: dict[str, dict] = {}
    tickers = yf.Tickers(" ".join(yf_symbols))
    for orig_sym, yf_sym in sym_map.items():
        try:
            hist = tickers.tickers[yf_sym].history(period="5d")
            if hist.empty:
                continue
            out[orig_sym] = {
                "price": float(hist["Close"].iloc[-1]),
                "price_date": hist.index[-1].date(),
            }
        except Exception as exc:  # network / delisted / bad symbol
            log.warning("price fetch failed for %s: %s", orig_sym, exc)
    return out

File: mobile/src/services/prices.ts
/**
 * Live prices.
 *
 * Mutual funds: AMFI publishes all NAVs daily as a semicolon-delimited text
 *               file at portal.amfiindia.com. Parsed by ISIN.
 * Stocks:       Yahoo Finance v8 chart API via plain fetch — no npm package.
 *               Symbols without a exchange suffix get .NS appended for NSE.
 */

const AMFI_NAV_URL = 'https://portal.amfiindia.com/spages/NAVAll.txt';
const AMFI_TTL_MS = 6 * 3600 * 1000;

interface AmfiRecord {
  nav: number;
  name: string;
  amfi_code: string;
  nav_date: string; // YYYY-MM-DD
}

let amfiCache: { fetchedAt: number; byIsin: Record<string, AmfiRecord> } = {
  fetchedAt: 0,
  byIsin: {},
};

/**
 * Fetch all AMFI NAVs and return a map keyed by ISIN.
 * Results are cached for 6 hours.
 */
export async function fetchAmfiNavs(force = false): Promise<Record<string, AmfiRecord>> {
  const now = Date.now();
  if (!force && Object.keys(amfiCache.byIsin).length && now - amfiCache.fetchedAt < AMFI_TTL_MS) {
    return amfiCache.byIsin;
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 20000);
  const resp = await fetch(AMFI_NAV_URL, { signal: controller.signal as any });
  clearTimeout(timeout);
  if (!resp.ok) throw new Error(`AMFI HTTP ${resp.status}`);
  const text = await resp.text();

  const byIsin: Record<string, AmfiRecord> = {};

  for (const line of text.split('\n')) {
    const parts = line.split(';');
    if (parts.length !== 6 || !/^\d+$/.test(parts[0].trim())) continue;
    const [code, isin1, isin2, name, navStr, dateStr] = parts.map((p: string) => p.trim());
    const nav = parseFloat(navStr);
    if (isNaN(nav)) continue;
    // Parse "dd-Mon-yyyy" date format
    const [d, m, y] = dateStr.split('-');
    const months: Record<string, string> = {
      jan: '01', feb: '02', mar: '03', apr: '04', may: '05', jun: '06',
      jul: '07', aug: '08', sep: '09', oct: '10', nov: '11', dec: '12',
    };
    const navDate = `${y}-${months[m?.toLowerCase()] || '01'}-${d.padStart(2, '0')}`;

    const rec: AmfiRecord = { nav, name, amfi_code: code, nav_date: navDate };
    for (const isin of [isin1, isin2]) {
      if (isin && isin !== '-') {
        byIsin[isin] = rec;
      }
    }
  }

  amfiCache = { fetchedAt: now, byIsin };
  return byIsin;
}

/** Normalise Indian ticker for yahoo-finance2: add .NS if missing. */
function yfSym(sym: string): string {
  return sym.includes('.') ? sym : `${sym}.NS`;
}

interface StockQuote {
  price: number;
  price_date: string; // YYYY-MM-DD
}

/**
 * Fetch current stock prices via Yahoo Finance v8 chart API (no npm package).
 * Returns a map keyed by the original symbol (without .NS suffix).
 */
export async function fetchStockPrices(symbols: string[]): Promise<Record<string, StockQuote>> {
  if (!symbols.length) return {};

  const symMap: Record<string, string> = {};
  for (const s of symbols) {
    symMap[s] = s.includes('.') ? s : `${s}.NS`;
  }

  const out: Record<string, StockQuote> = {};
  for (const [origSym, yfSymStr] of Object.entries(symMap)) {
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 10000);
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(yfSymStr)}?interval=1d&range=5d`;
      const resp = await fetch(url, {
        signal: controller.signal as any,
        headers: { 'User-Agent': 'Mozilla/5.0' },
      });
      clearTimeout(timeout);
      if (!resp.ok) continue;
      const json: any = await resp.json();
      const meta = json?.chart?.result?.[0]?.meta;
      if (!meta) continue;
      const price = meta.regularMarketPrice;
      if (price == null) continue;
      const ts = meta.regularMarketTime
        ? new Date(meta.regularMarketTime * 1000)
        : new Date();
      out[origSym] = {
        price: typeof price === 'number' ? price : parseFloat(String(price)),
        price_date: ts.toISOString().slice(0, 10),
      };
    } catch {
      // symbol not found / delisted — skip silently
    }
  }

  return out;
}
